### DISCUSSION TITLE
Having another problem - 'Cannot call unknown function', but the function exists

### DISCUSSION THREAD

**Discussion by inhahe** [emscripten-core/emscripten#22974] [Q&A]
Here's my HTML:
```
<!DOCTYPE html>
<html>
  <head>
      <meta charset="utf-8">
      <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
  </head>
  <body>
    <!-- Create the canvas that the C++ code will draw into -->
    <canvas id="canvas" oncontextmenu="event.preventDefault()"></canvas>

    <!-- Allow the C++ to access the canvas element -->
    <script type='text/javascript'>
      var Module = {
        canvas: (function () { return document.getElementById('canvas'); })()
      };
    </script>

    <script>
      Module.onRuntimeInitialized = () =>
      {
        console.log(...Object.keys(Module)); // Lists all exported functions and variables
        const pixelsptr = Module.ccall(_getpixels, 'number', [], []);
      };
    </script>

    <!-- Add the javascript glue code (index.js) as generated by Emscripten -->
    <script src="em.js"></script>

  </body>
</html>
```
Here's the relevant C++ lines:
```
static uint8_t pixels[w * h * 4];
extern "C" {uint8_t* EMSCRIPTEN_KEEPALIVE getpixels() { return &pixels[0]; }}
```
Here's my em++ command line:
`em++ /mnt/d/visual\ studio\ projects/scribbles3/em.cpp -s WASM=1 -s -o em.js -sALLOW_MEMORY_GROWTH -Oz -sEXPORTED_FUNCTIONS=_getpixels,_mainloop,_main -sEXPORTED_RUNTIME_METHODS=ccall,cwrap -s ASSERTIONS=2`
Here's my console output:
```anvas onRuntimeInitialized FS_createDataFile FS_createPreloadedFile _getpixels _mainloop _update _main ___set_stack_limits dynCall_jiji ccall cwrap HEAP8 HEAP16 HEAPU8 HEAPU16 HEAP32 HEAPU32 HEAPF32 HEAPF64 calledRun
em.js:1 Aborted(Assertion failed: Cannot call unknown function (...args)=>{assert(runtimeInitialized,`native function \`${name}\` called before runtime initialization`);var f=wasmExports[name];assert(f,`exported native function \`${name}\` not found`);assert(args.length<=nargs,`native function \`${name}\` called with ${args.length} args but expects ${nargs}`);return f(...args)}, make sure it is exported)
abort @ em.js:1
em.js:1 Uncaught (in promise) RuntimeError: Aborted(Assertion failed: Cannot call unknown function (...args)=>{assert(runtimeInitialized,`native function \`${name}\` called before runtime initialization`);var f=wasmExports[name];assert(f,`exported native function \`${name}\` not found`);assert(args.length<=nargs,`native function \`${name}\` called with ${args.length} args but expects ${nargs}`);return f(...args)}, make sure it is exported)
    at abort (em.js:1:10513)
    at assert (em.js:1:5912)
    at getCFunc (em.js:1:26423)
    at Object.ccall (em.js:1:27085)
    at Module.onRuntimeInitialized (monkey_test.html:22:34)
    at doRun (em.js:1:35873)
    at run (em.js:1:36050)
    at runCaller (em.js:1:34923)
    at removeRunDependency (em.js:1:10401)
    at receiveInstance (em.js:1:13705)
```
As you can see, '_getpixels' is actaully listed in the module's contents, and it also shouldn't run before it's initialized, despite what it says, because it runs under `onRuntimeInitialized'.

**Comment by sbc100** [comment]
The ccall function expects a string as its first argument, not a function.   So you would do `ccall('getpixels')` (also note that you don't need that leading underscore here).

Also, if you do `EMSCRIPTEN_KEEPALIVE` you don't *also* need to add it to `EXPORTED_FUNCTIONS` (those two things have the same effect).

**Reply by inhahe** [reply]
> The ccall function expects a string as its first argument, not a function. So you would do `ccall('getpixels')` (also note that you don't need that leading underscore here).
> 
> Also, if you do `EMSCRIPTEN_KEEPALIVE` you don't _also_ need to add it to `EXPORTED_FUNCTIONS` (those two things have the same effect).

Thanks, I'll try that. For some reason the docs didn't seem to explain or give an example of how to use ccall, that I remember.

Regarding keepalive and exported_functions, I had just the EMSCRIPTEN_KEEPALIVE and it didn't export them. I guess you mean EXPORTED_FUNCTIONS also implies EMSCRIPTEN_KEEPALIVE but not vice versa

**Reply by sbc100** [reply]
`EMSCRIPTEN_KEEPALIVE` implies `EXPORTED_FUNCTIONS`.. thats really all it does

**Reply by sbc100** [reply]
The docs for ccall and cwap do you strings: https://emscripten.org/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html#calling-compiled-c-functions-from-javascript-using-ccall-cwrap.  Although maybe we should add more assertions if its confusing.

**Reply by inhahe** [reply]
> The docs for ccall and cwap do you strings: https://emscripten.org/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html#calling-compiled-c-functions-from-javascript-using-ccall-cwrap. Although maybe we should add more assertions if its confusing.

Oh, yes, I see it does say how to use ccall. I guess I just overlooked the first parameter! duh. thanks. 

I guess I should have known '_getpixels' wouldn't be in the global namespace, i must not have noticed i forgot the quotes.
